library/core/src/mem/manually_drop.rs

```diff
@@ -40,9 +39,114 @@ use crate::ptr;
 /// }
 /// ```
 ///
+/// # Interaction with `Box`.
+///
+/// Currently, once the `Box<T>` inside a `ManuallyDrop<Box<T>>` is dropped,
+/// moving the `ManuallyDrop<Box<T>>` is [considered to be undefined
+/// behavior](https://github.com/rust-lang/unsafe-code-guidelines/issues/245).
+/// That is, the following code causes undefined behavior:
+///
+/// ```no_run
+/// use std::mem::ManuallyDrop;
+///
+/// let mut x = ManuallyDrop::new(Box::new(42));
+/// unsafe {
+///     ManuallyDrop::drop(&mut x);
+/// }
+/// let y = x; // Undefined behavior!
+/// ```
+///
+/// This may change in the future. In the meantime, consider using
+/// [`MaybeUninit`] instead.
+///
+/// # Safety hazards when storing `ManuallyDrop` in a struct / enum.
+///
+/// Special care is needed when all of the conditions below are met:
+/// * A field of a struct or enum is a `ManuallyDrop` or contains a
+///   `ManuallyDrop`, without the `ManuallyDrop` being inside a `union`.
+/// * The struct or enum is part of public API, or is stored in a struct or enum
+///   that is part of public API.
+/// * There is code outside of a `Drop` implementation that calls
+///   [`ManuallyDrop::drop`] or [`ManuallyDrop::take`] on the `ManuallyDrop`
+///   field.
+///
+/// In particular, the following hazards can occur:
+///
+/// #### Storing generic types
+///
+/// If the `ManuallyDrop` contains a client-supplied generic type, the client
+/// might provide a `Box`, causing undefined behavior when the struct / enum is
+/// later moved, as mentioned above. For example, the following code causes
+/// undefined behavior:
+///
+/// ```no_run
+/// use std::mem::ManuallyDrop;
+///
+/// pub struct BadOption<T> {
+///     // Invariant: Has been dropped iff `is_some` is false.
+///     value: ManuallyDrop<T>,
+///     is_some: bool,
+/// }
+/// impl<T> BadOption<T> {
+///     pub fn new(value: T) -> Self {
+///         Self { value: ManuallyDrop::new(value), is_some: true }
+///     }
+///     pub fn change_to_none(&mut self) {
+///         if self.is_some {
+///             self.is_some = false;
+///             unsafe {
+///                 // SAFETY: `value` hasn't been dropped yet, as per the invariant
+///                 // (This is actually unsound!)
+///                 ManuallyDrop::drop(&mut self.value);
+///             }
+///         }
+///     }
+/// }
+///
+/// // In another crate:
+///
+/// let mut option = BadOption::new(Box::new(42));
+/// option.change_to_none();
+/// let option2 = option; // Undefined behavior!
+/// ```
+///
+/// #### Deriving traits
+///
+/// Deriving `Debug`, `Clone`, `PartialEq`, `PartialOrd`, `Ord`, or `Hash` on
+/// the struct / enum could be unsound, since the derived implementations of
+/// these traits would access the `ManuallyDrop` field. For example, the
+/// following code causes undefined behavior:
+///
+/// ```no_run
+/// use std::mem::ManuallyDrop;
+///
+/// #[derive(Debug)] // This is unsound!
```

Comment:
I'm not sure what's the best way to phrase this, but I don't think "unsound" is right here. 

The drive on its own *is* sound. It's the combination with `ManuallyDrop::drop` which is unsound. And since the latter is unsafe, it falls into its safety requirements that the `ManuallyDrop` value is not used after the call. 

That being said, basically all intended usages of `ManuallyDrop` do not combine with derived traits (since the user of the type can always call them after the value is dropped), so this is an important foot gun to highlight. I just feel like there is a better way to explain it maybe.

Reply:
I've changed the comment to: "This derive is unsound in combination with the `ManuallyDrop::drop` call."